README.rst
Disk Patrol
===========

This is a program to check disk space and alert
if the used space is above the configured threshold.

Here are the alert transports.

* Wall transport: send alert message to the logged-in users.
* Email transport: send alert message to the mail recipients.
* Telegram transport: send alert message to the telegram chat id.
* Slack transport: send alert message to the slack chaneel.

Build (Optional)
-----------------

If you want to build rpm package, install docker and run run.sh script.::

    $ ./run.sh -h
    USAGE: run.sh [-h] [-b] [-r] [debian|rocky|centos7]
    
     -h                             Display this help message.
     -b [debian|rocky|centos7]      Build diskpatrol.
     -r [debian|rocky|centos7]      Run into diskpatrol build env.

To build rpm package for rocky linux::

    $ ./run.sh -b rocky

Install
--------

For RHEL-based distro
+++++++++++++++++++++++

Install diskpatrol rpm package from the internal genesis file server.::

    $ sudo dnf -y install \
        http://192.168.151.110:8000/burrito/diskpatrol-0.2.0-1.el8.x86_64.rpm

Optionally, Install postfix rpm package from the internal genesis file server
if you want to use email transport.::

    $ sudo dnf -y install \
        http://192.168.151.110:8000/burrito/postfix-3.5.8-4.el8.x86_64.rpm

For Debian-based distro
++++++++++++++++++++++++

Install diskpatrol debian package from the internal genesis file server.::

    $ curl -LO http://192.168.151.110:8000/burrito/diskpatrol_0.2.0-1_amd64.deb
    $ sudo dpkg -i diskpatrol_0.2.0-1_amd64.deb

Optionally, Install postfix rpm package if you want to use email transport.::

    $ sudo apt update && sudo apt install postfix

Choose 'No configuration' when it asks the mail server configuration type.


Configure
----------

If there is /etc/postfix/main.cf, back it up to /etc/postfix/main.cf.bak.::

    $ sudo mv /etc/postfix/main.cf /etc/postfix/main.cf.bak

Edit /etc/postfix/main.cf.::

    $ sudo vi /etc/postfix/main.cf
    inet_interfaces = loopback-only
    myorigin = <host_fqdn>
    mydestination = 
    local_transport = error: local delivery disabled
    mynetworks = 127.0.0.0/8,[::1]/128
    smtpd_relay_restrictions = permit_mynetworks,reject

Restart postfix service.::

    $ sudo systemctl restart postfix.service

Edit /etc/diskpatrol.conf.

The configuration is self-explanatory. Read the comment and set the
configuration variables for your environment.

Restart diskpatrol service.::

    $ sudo systemctl restart diskpatrol.service

Operations
-----------

Check diskpatrol service is running.::

    $ sudo systemctl status diskpatrol.service

The diskpatrol log file is /var/log/diskpatrol.log
Check the diskpatrol log file.::

    $ tail -f /var/log/diskpatrol.log
    2023-08-30 21:29:53,956 INFO:The used space of / (33%) is below warning level (60%)
    2023-08-30 21:30:54,026 INFO:The used space of / (33%) is below warning level (60%)
    2023-08-30 21:31:54,060 INFO:The used space of / (33%) is below warning level (60%)
    2023-08-30 21:32:54,122 INFO:The used space of / (33%) is below warning level (60%)

If the used space is above one of thresholds, it will alert like this.

* Wall alert::

    DiskPatrol: / is above warning level.

    The used space of / is above warning level (30%). 
    used/total: 32.6GiB/100.0GiB (33% used)
                                                                           
    Please delete some files in /.

* Email alert::

    Date: Thu, 31 Aug 2023 16:37:35
    From: diskpatrol@bbeta-controller.cluster.local
    Subject: DiskPatrol: / is above warning level.
    
    The used space of / is above warning level (30%).
    used/total: 32.6GiB/100.0GiB (33% used)
    
    Please delete some files in /.

* Telegram alert::

    diskpatrol@bbeta-controller
    DiskPatrol: / is in warning level.
    
    The used space of / is above warning level (30%).
    used/total: 32.6GiB/100.0GiB (33% used)
    
    Please delete some files in /.

To add telegram alert, create a telegram bot.
Here is the guide how to create a telegram bot.

https://core.telegram.org/bots/features#botfather

You need to set APITOKEN and CHATID in /etc/diskpatrol.conf to send
telegram message.



debian/changelog
diskpatrol (0.2.1-1) unstable; urgency=medium

  * Fix: add strip for each path in PATHS variable

 -- Heechul Kim <jijisa@iorchard.net>  Fri, 15 Sep 2023 18:18:57 +0900

diskpatrol (0.2.0-1) unstable; urgency=medium

  * Initial release

 -- Heechul Kim <jijisa@iorchard.net>  Thu, 14 Sep 2023 08:34:50 +0000


run.sh
#!/bin/bash

set -e

function USAGE() {
  echo "USAGE: $(basename $0) [-h] [-b] [-r] [debian|rocky]" 1>&2
  echo
  echo " -h                             Display this help message."
  echo " -b [debian|rocky|centos7]      Build diskpatrol."
  echo " -r [debian|rocky|centos7]      Run into diskpatrol build env."
  echo
}
while getopts 'b:r:h' opt; do
  case "${opt}" in
    b)
      DISTRO="${OPTARG}"
      break
      ;;
    r)
      DISTRO="${OPTARG}"
      break
      ;;
    h)
      USAGE
      exit 0
      ;;
    *)
      exit 1
      ;;
  esac
done

case "${DISTRO}" in
  centos7)
    FROM="docker.io/centos:7"
    ;;
  debian)
    FROM="docker.io/debian:10-slim"
    ;;
  rocky)
    FROM="docker.io/rockylinux/rockylinux:8"
    ;;
  *)
    echo "Unknown distro ${DISTRO}."
    USAGE
    exit 1
    ;;
esac

IMAGE="docker.io/jijisa/diskpatrol-builder:${DISTRO}"

docker build -t ${IMAGE} --build-arg FROM=${FROM} -f Dockerfile .
if [[ "$opt" = "b" ]]; then
  docker run -v $(pwd)/output:/output --rm ${IMAGE}
fi
if [[ "$opt" = "r" ]]; then
  docker run -it -v $(pwd)/output:/output --rm --entrypoint=/bin/bash ${IMAGE}
fi



scripts/build.sh
#!/bin/bash

set -ex

PKGDIR="${WORKSPACE}/pkgdir"
BINFILE="${PKGDIR}/build/x86_64-unknown-linux-gnu/debug/exe/diskpatrol"

function get_distro() {
  if [[ -f /etc/os-release ]]; then
    source /etc/os-release
    echo ${ID}-${VERSION_ID}
  else
    echo "Unknown"
  fi
}

function prepare_centos7() {
  yum -y install python3-pip curl gcc make rpmdevtools rpmlint
}
function prepare_rocky() {
  dnf -y install python3-pip curl gcc make rpmdevtools rpmlint
}
function prepare_debian() {
  apt update
  DEBIAN_FRONTEND=noninteractive apt -y install python3-pip curl gcc make \
    dpkg-dev debhelper
}

function build_deb() {
  # clean /diskpatrol-${VER}
  rm -fr /diskpatrol-${VER}
  # copy /build to /diskpatrol-${VER}
  cp -a /build /diskpatrol-${VER}
  # copy diskpatrol.service to debian dir
  cp /build/scripts/diskpatrol.service /diskpatrol-${VER}/debian/
  # create a tarball
  pushd /
    tar cvzf diskpatrol_${VER}.orig.tar.gz diskpatrol-${VER}
  popd
  # build a debian package
  pushd /diskpatrol-${VER}
    dpkg-buildpackage
  popd
  cp /diskpatrol_${VER}*.deb ${OUTPUT_DIR}
}
function build_deb_from_binary() {
  DEBBUILD="/diskpatrol_${VER}-1_amd64"
  mkdir -p ${DEBBUILD}/{DEBIAN,usr/bin,etc/logrotate.d,etc/systemd/system}
  pushd ${WORKSPACE}/scripts
    sed -i "s/%%VERSION%%/${VER}-1/" DEBIAN/control
    cp -a DEBIAN ${DEBBUILD}/
    cp diskpatrol.conf.sample ${DEBBUILD}/etc/diskpatrol.conf
    cp diskpatrol.service ${DEBBUILD}/etc/systemd/system/
    cp diskpatrol.logrotate ${DEBBUILD}/etc/logrotate.d/diskpatrol
    cp ${BINFILE} ${DEBBUILD}/usr/bin/
  popd
  pushd /
    dpkg-deb --build --root-owner-group ${DEBBUILD}
  popd
  cp ${DEBBUILD}.deb ${OUTPUT_DIR}
}
function build_rpm() {
  python3 -m pip install -U pip
  python3 -m pip install pyoxidizer
  curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | \
    sh -s -- --no-modify-path -y

  # build diskpatrol binary
  pyoxidizer build --system-rust --path=${PKGDIR}
  strip ${BINFILE}
  chmod 0755 ${BINFILE}
  cp ${BINFILE} ${OUTPUT_DIR}

  rpmdev-setuptree
  pushd ${WORKSPACE}/scripts
    sed -i "s/%%VERSION%%/${VER}/" diskpatrol.spec
    cp diskpatrol.spec /root/rpmbuild/SPECS/
    mkdir -p /diskpatrol-${VER}
    cp diskpatrol.conf.sample /diskpatrol-${VER}/diskpatrol.conf
    cp diskpatrol.service ${BINFILE} /diskpatrol-${VER}/
    cp diskpatrol.logrotate /diskpatrol-${VER}/
  popd
  pushd /
    tar cvzf /root/rpmbuild/SOURCES/diskpatrol-${VER}.tar.gz diskpatrol-${VER}
  popd
  rpmbuild -bb /root/rpmbuild/SPECS/diskpatrol.spec
  cp /root/rpmbuild/RPMS/x86_64/diskpatrol-${VER}*.rpm ${OUTPUT_DIR}
}

function main() {
  case $(get_distro) in
    centos-7)
      prepare_centos7
      ;;
    rocky-8.8)
      prepare_rocky
      ;;
    debian-10)
      prepare_debian
      ;;
  esac
  VER=$(python3 ${PKGDIR}/diskpatrol/__init__.py)

  case $(get_distro) in
    centos-7)
      build_rpm
      ;;
    rocky-8.8)
      build_rpm
      ;;
    debian-10)
      build_deb
      ;;
  esac
}

main



diskalert/diskalert.py
import os
import sys
import math
import time
import shutil
import logging
import argparse

from subprocess import Popen

from dotenv import dotenv_values


def main(d_vars: dict) -> None:
    """main program"""
    try:
        logging.basicConfig(filename=d_vars['logfile'], level=logging.INFO)
    except Exception as e:
        logging.exception(e)
        parser.exit(1, message=str(e))

    if not os.path.isfile(d_vars['envfile']):
        s_msg = f"Error: cannot find the env file {d_vars['envfile']}\n"
        logging.error(s_msg)
        parser.exit(1, s_msg)

    config = {
        **dotenv_values(d_vars['envfile']),
    }
    
    d_thresholds = {
        'warn': int(config['WARN']),
        'error': int(config['ERROR']),
        'critical': int(config['CRITICAL']),
    }
    
    while True:
        o_stat = shutil.disk_usage(config['PARTITION'])
        i_used_percent = int(math.ceil(o_stat.used/o_stat.total*100))
        if i_used_percent > d_thresholds['warn']:
            if i_used_percent > d_thresholds['critical']:
                s_level = 'critical'
            elif i_used_percent > d_thresholds['error']:
                s_level = 'error'
            else:
                s_level = 'warn'
            s_msg_file = config['MSGDIR'] + '/diskalert_' + s_level
            s_msg = f'''
            The used space of root FS is above {s_level} threshold 
            ({d_thresholds[s_level]}%).
            Please delete some files in root FS.
            '''
            with open(s_msg_file, 'w') as f:
                f.write(s_msg)
            s_wall_cmd = f'cat {s_msg_file} | wall'
            o_proc = Popen(s_wall_cmd, shell=True)
            print(o_proc.returncode)
        else:
            # remove a message file
            if os.path.isfile(s_msg_file):
                os.unlink(s_msg_file)
        time.sleep(int(config['SLEEP']))

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
            prog='diskalert',
            allow_abbrev=False,
            description='Alert when disk space is lower than thresholds',
            epilog='Thanks for using %(prog)s! :)',
            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('-d', '--debug', action='store_true',
            help='enable debug mode')
    parser.add_argument('-l', '--logfile', default='/var/log/diskalert.log',
            help='path to logfile')
    parser.add_argument('envfile', help='environment file location')
    args = parser.parse_args()
    d_vars = vars(args)
    print(d_vars)
    #i_level = getattr(logging, loglevel.upper(), None)
    main(d_vars)


diskpatrol/diskpatrol.py
import os
import sys
import math
import time
import shutil
import signal
import logging
import argparse

from subprocess import Popen

from dotenv import dotenv_values

def handler(signum, frame):
    res = input("Ctrl-C is pressed. Do you want to exit? (y/n) ")
    if res == 'y':
        sys.exit(0)

signal.signal(signal.SIGINT, handler)

def app(d_vars: dict) -> None:
    """diskalert application"""
    try:
        o_level = logging.DEBUG if d_vars['debug'] else logging.INFO
        logging.basicConfig(
            format='%(asctime)s %(message)s',
            filename=d_vars['logfile'],
            level=o_level)
    except Exception as e:
        logging.exception(e)
        parser.exit(1, message=str(e))

    if not os.path.isfile(d_vars['envfile']):
        s_msg = f"Error: cannot find the env file {d_vars['envfile']}\n"
        logging.error(s_msg)
        parser.exit(1, s_msg)

    config = {
        **dotenv_values(d_vars['envfile']),
    }
    
    d_thresholds = {
        'warn': int(config['WARN']),
        'error': int(config['ERROR']),
        'critical': int(config['CRITICAL']),
    }
    
    while True:
        o_stat = shutil.disk_usage(config['PARTITION'])
        i_used_percent = int(math.ceil(o_stat.used/o_stat.total*100))
        if i_used_percent > d_thresholds['warn']:
            if i_used_percent > d_thresholds['critical']:
                s_level = 'critical'
            elif i_used_percent > d_thresholds['error']:
                s_level = 'error'
            else:
                s_level = 'warn'
            s_msg_file = config['MSGDIR'] + '/diskalert_' + s_level
            s_msg = f'''
            The used space of root FS is above {s_level} threshold 
            ({d_thresholds[s_level]}%).
            Please delete some files in root FS.
            '''
            with open(s_msg_file, 'w') as f:
                f.write(s_msg)
            s_wall_cmd = f'cat {s_msg_file} | wall'
            o_proc = Popen(s_wall_cmd, shell=True)
        else:
            # remove a message file
            if os.path.isfile(s_msg_file):
                os.unlink(s_msg_file)
        time.sleep(int(config['SLEEP']))

def main():
    """Diskalert main program"""
    parser = argparse.ArgumentParser(
            prog='diskalert',
            allow_abbrev=False,
            description='Alert when disk used space is higher than thresholds',
            epilog='Thanks for using %(prog)s! :)',
            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('-d', '--debug', action='store_true',
            help='enable debug mode')
    parser.add_argument('-l', '--logfile', default='/var/log/diskalert.log',
            help='path to logfile')
    parser.add_argument('envfile', help='environment file location')
    args = parser.parse_args()
    d_vars = vars(args)
    print(d_vars)
    app(d_vars)


pkgdir/diskpatrol/cli.py
import os
import sys
import glob
import math
import time
import shutil
import signal
import socket
import logging
import smtplib
import argparse
import requests

from subprocess import Popen
from dotenv import dotenv_values

MSGDIR = "/tmp"
PROGNAME = 'diskpatrol'

d_level = { "debug": 10, "info": 20, "warning": 30, 
            "error": 40, "critical": 50 }

def clean_tmp():
    try:
        logging.debug(f"clean up diskpatrol temp files")
        for f in glob.glob(f"{MSGDIR}/{PROGNAME}-*"):
            os.unlink(f)
            logging.debug(f"deleted {f}")
    except Exception as e:
        logging.exception(e)

def handler(signum, frame):
    if signum == signal.SIGINT:
        res = input("Ctrl-C is pressed. Do you want to exit? (y/n) ")
        if res == 'y':
            sys.exit(0)
    if signum == signal.SIGHUP:
        logging.info("Received SIGHUP. Reload config file.")
        pass

signal.signal(signal.SIGINT, handler)
signal.signal(signal.SIGHUP, handler)

def send_telegram(msg: str, token: str, chatid: str) -> None:
    """Send a Telegram message to a chat id."""
    s_url = f"https://api.telegram.org/bot{token}/sendMessage"
    logging.debug(s_url)
    try:
        response = requests.post(s_url, json={'chat_id': chatid, 'text': msg})
        logging.debug(response.text)
        logging.debug("Sent telegram alert.")
    except Exception as e:
        logging.error(e)

def send_slack(msg: str, webhook: str) -> requests.models.Response:
    """Send a Slack message to a channel via a webhook."""
    try:
        response = requests.post(webhook, json={'text': msg})
        logging.debug(response.text)
        logging.debug("Sent slack alert.")
    except Exception as e:
        logging.error(e)

def humanize(num, suffix="B"):
    for unit in ("", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"):
        if abs(num) < 1024.0:
            return f"{num:3.1f}{unit}{suffix}"
        num /= 1024.0
    return f"{num:.1f}Yi{suffix}"

def app(d_vars: dict, config: dict) -> None:
    """diskpatrol application"""
    s_fqdn = socket.getaddrinfo(socket.gethostname(), 0, 
            flags=socket.AI_CANONNAME)[0][3]
    s_sender = f'diskpatrol@{s_fqdn}'
    l_recipients = config['RECIPIENT_EMAILS'].split(',')

    d_thresholds = {
        'warning': int(config['WARNING']),
        'error': int(config['ERROR']),
        'critical': int(config['CRITICAL']),
    }
    
    while True:
        for s_path in [s.strip() for s in config['PATHS'].split(',')]:
            s_msgfile_prefix = MSGDIR + '/' + PROGNAME
            if os.path.basename(s_path):
                s_msgfile_prefix += s_path.replace('/', '-')
            else:
                s_msgfile_prefix += '-root'
            o_stat = shutil.disk_usage(s_path)
            i_used_percent = int(math.ceil(o_stat.used/o_stat.total*100))
            if i_used_percent >= d_thresholds['warning']:
                if i_used_percent >= d_thresholds['critical']:
                    s_level = 'critical'
                elif i_used_percent >= d_thresholds['error']:
                    s_level = 'error'
                else:
                    s_level = 'warning'
                s_msgfile = s_msgfile_prefix + '_' + s_level
                s_msg = f"The used space of {s_path} is above {s_level} " + \
                        f"level ({d_thresholds[s_level]}%).\n" + \
                        "used/total: " + \
                        f"{humanize(o_stat.used)}/{humanize(o_stat.total)}" + \
                        f" ({i_used_percent}% used)" + \
                        "\n\n" + \
                        f"Please delete some files in {s_path}."
                s_subject = f"DiskPatrol: {s_path} is above {s_level} level."
                s_log = s_msg.replace('\n', ' ')
                if s_level == 'critical':
                    logging.critical(s_log)
                elif s_level == 'error':
                    logging.error(s_log)
                else:
                    logging.warning(s_log)
                with open(s_msgfile, 'w') as f:
                    f.write(f"{s_subject}\n\n{s_msg}")
                if config['ENABLE_WALL'] == '1' and \
                    d_level[s_level] >= d_level[config['WALL_ALERT_LEVEL']]:
                    logging.debug("Send wall alert.")
                    s_wall_cmd = f"cat {s_msgfile} | wall -n"
                    o_proc = Popen(s_wall_cmd, shell=True)
                if config['ENABLE_EMAIL'] == '1' and \
                    d_level[s_level] >= d_level[config['EMAIL_ALERT_LEVEL']]:
                    s_emailcount = f"{s_msgfile_prefix}_{s_level}_emailcount"
                    if os.path.isfile(s_emailcount):
                        with open(s_emailcount, 'r') as f:
                            i_emailcount = int(f.read())
                    else:
                        # remove other level count files when level is changed.
                        logging.debug(f"Alert level is changed to {s_level}")
                        logging.debug("Remove other level count files.")
                        if s_level == 'critical':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_emailcount",
                                f"{s_msgfile_prefix}_error_emailcount"
                            ]
                        elif s_level == 'error':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_emailcount",
                                f"{s_msgfile_prefix}_critical_emailcount"
                            ]
                        else:
                            l_countfile = [
                                f"{s_msgfile_prefix}_error_emailcount",
                                f"{s_msgfile_prefix}_critical_emailcount"
                            ]
                        for f in l_countfile:
                            if os.path.isfile(f):
                                os.unlink(f)
                        logging.debug(f"Create {s_emailcount}.")
                        with open(s_emailcount, 'w') as f:
                            f.write('1')
                        i_emailcount = 1
                    if i_emailcount > \
                        int(config['EMAIL_ALERT_COUNT']) + \
                        int(config['EMAIL_ALERT_SKIP']):
                        logging.debug("Reset the email alert count.")
                        with open(s_emailcount, 'w') as f:
                            f.write('1')
                        i_emailcount = 1
                    if i_emailcount <= int(config['EMAIL_ALERT_COUNT']):
                        logging.debug(f"Email alert {i_emailcount} " + \
                                    f"out of {config['EMAIL_ALERT_COUNT']}")
                        logging.debug("Send email alert.")
                        s_mail = f"From: {s_sender}\n" + \
                                f"To: {config['RECIPIENT_EMAILS']}\n" + \
                                f"Subject: {s_subject}\n\n" + \
                                s_msg
                        with smtplib.SMTP(host=config['SMTP_SERVER'],
                                port=int(config['SMTP_PORT'])) as smtp:
                            smtp.sendmail(s_sender, l_recipients, s_mail)
                    else:
                        logging.debug("Skipping Email alert.")
                    with open(s_emailcount, 'w') as f:
                        f.write(str(i_emailcount+1))
                if config['ENABLE_TELEGRAM'] == '1' and \
                    d_level[s_level]>=d_level[config['TELEGRAM_ALERT_LEVEL']]:
                    s_telegramcount = f"{s_msgfile_prefix}_{s_level}_" + \
                                        "telegramcount"
                    if os.path.isfile(s_telegramcount):
                        with open(s_telegramcount, 'r') as f:
                            i_telegramcount = int(f.read())
                    else:
                        # remove other level count files when level is changed.
                        logging.debug(f"Alert level is changed to {s_level}")
                        logging.debug("Remove other level count files.")
                        if s_level == 'critical':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_telegramcount",
                                f"{s_msgfile_prefix}_error_telegramcount"
                            ]
                        elif s_level == 'error':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_telegramcount",
                                f"{s_msgfile_prefix}_critical_telegramcount"
                            ]
                        else:
                            l_countfile = [
                                f"{s_msgfile_prefix}_error_telegramcount",
                                f"{s_msgfile_prefix}_critical_telegramcount"
                            ]
                        for f in l_countfile:
                            if os.path.isfile(f):
                                os.unlink(f)
                        logging.debug(f"Create {s_telegramcount}.")
                        with open(s_telegramcount, 'w') as f:
                            f.write('1')
                        i_telegramcount = 1
                    if i_telegramcount > \
                        int(config['TELEGRAM_ALERT_COUNT']) + \
                        int(config['TELEGRAM_ALERT_SKIP']):
                        logging.debug("Reset the telegram alert count.")
                        with open(s_telegramcount, 'w') as f:
                            f.write('1')
                        i_telegramcount = 1
                    if i_telegramcount <= int(config['TELEGRAM_ALERT_COUNT']):
                        logging.debug(f"Telegram alert: {i_telegramcount} " + \
                                  f"out of {config['TELEGRAM_ALERT_COUNT']}")
                        logging.debug("Send telegram alert.")
                        if config['APITOKEN'] and config['CHATID']:
                            s_telegram = f"{s_sender}\n{s_subject}\n\n{s_msg}"
                            send_telegram(s_telegram, config['APITOKEN'],
                                config['CHATID'])
                        else:
                            logging.warning("APITOKEN and/or CHATID not set")
                    else:
                        logging.debug("Skipping Telegram alert.")
                    with open(s_telegramcount, 'w') as f:
                        f.write(str(i_telegramcount+1))
                if config['ENABLE_SLACK'] == '1' and \
                    d_level[s_level]>=d_level[config['SLACK_ALERT_LEVEL']]:
                    s_slackcount = f"{s_msgfile_prefix}_{s_level}_" + \
                                        "slackcount"
                    if os.path.isfile(s_slackcount):
                        with open(s_slackcount, 'r') as f:
                            i_slackcount = int(f.read())
                    else:
                        logging.debug(f"Alert level is changed to {s_level}")
                        logging.debug("Remove other level count files.")
                        if s_level == 'critical':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_slackcount",
                                f"{s_msgfile_prefix}_error_slackcount"
                            ]
                        elif s_level == 'error':
                            l_countfile = [
                                f"{s_msgfile_prefix}_warning_slackcount",
                                f"{s_msgfile_prefix}_critical_slackcount"
                            ]
                        else:
                            l_countfile = [
                                f"{s_msgfile_prefix}_error_slackcount",
                                f"{s_msgfile_prefix}_critical_slackcount"
                            ]
                        for f in l_countfile:
                            if os.path.isfile(f):
                                os.unlink(f)
                        logging.debug(f"Create {s_slackcount}.")
                        with open(s_slackcount, 'w') as f:
                            f.write('1')
                        i_slackcount = 1
                    if i_slackcount > \
                        int(config['SLACK_ALERT_COUNT']) + \
                        int(config['SLACK_ALERT_SKIP']):
                        logging.debug("Reset the slack alert count.")
                        with open(s_slackcount, 'w') as f:
                            f.write('1')
                        i_slackcount = 1
                    if i_slackcount <= int(config['SLACK_ALERT_COUNT']):
                        logging.debug(f"Slack alert: {i_slackcount} " + \
                                  f"out of {config['SLACK_ALERT_COUNT']}")
                        logging.debug("Send slack alert.")
                        if config['SLACK_WEBHOOK']:
                            s_slack = f"{s_sender}\n{s_subject}\n\n{s_msg}"
                            send_slack(s_slack, config['SLACK_WEBHOOK'])
                        else:
                            logging.warning("SLACK_WEBHOOK is not set")
                    else:
                        logging.debug("Skipping Slack alert.")
                    with open(s_slackcount, 'w') as f:
                        f.write(str(i_slackcount+1))
            else:
                s_log = f"The used space of {s_path} ({i_used_percent}%) " + \
                        f"is below warning level ({d_thresholds['warning']}%)"
                logging.info(s_log)
                # clean up message files
                logging.debug(f"clean up {s_msgfile_prefix}_*")
                try:
                    for f in glob.glob(f"{s_msgfile_prefix}_*"):
                        os.unlink(f)
                except Exception as e:
                    logging.exception(e)
        time.sleep(int(config['SLEEP']))

def main():
    """Disk Patrol main program"""
    parser = argparse.ArgumentParser(
            prog='diskpatrol',
            allow_abbrev=False,
            description='Alert when disk used space is higher than thresholds',
            epilog='Thanks for using %(prog)s! :)',
            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    #parser.add_argument('-v', '--version', action='version',
    #        version="{version}".format(version=__version__))
    parser.add_argument('-d', '--debug', action='store_true',
            help='enable debug mode')
    parser.add_argument('-l', '--logfile', help='path to logfile')
    #parser.add_argument('-c', '--conf', 
    #        default='/etc/diskpatrol.conf',
    #        help='configuration file location')
    parser.add_argument('conf', help='configuration file location')
    args = parser.parse_args()
    d_vars = vars(args)
    if not d_vars['conf'] or not os.path.isfile(d_vars['conf']):
        s_msg = f"Error: cannot find the conf file {d_vars['conf']}\n"
        logging.error(s_msg)
        parser.exit(1, s_msg)

    config = {
        **dotenv_values(d_vars['conf']),
    }
    level = logging.DEBUG if d_vars['debug'] else d_level[config['LOGLEVEL']]
    logfile = d_vars['logfile'] if d_vars['logfile'] else config['LOGFILE']
    try:
        logging.basicConfig(
            format='%(asctime)s %(levelname)s:%(message)s',
            filename=logfile,
            level=level)
    except Exception as e:
        logging.exception(e)
        parser.exit(1, message=str(e))
    clean_tmp()
    app(d_vars, config)

if __name__ == '__main__':
    main()
